flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/FlinkKafkaInternalProducer.java

```diff
@@ -396,8 +408,27 @@ public String toString() {
                 + transactionalId
                 + "', inTransaction="
                 + inTransaction
+                + ", pendingRecords="
+                + pendingRecords.get()
                 + ", closed="
                 + closed
                 + '}';
     }
+
+    public class TrackingCallback implements Callback {
+
+        private final Callback actualCallback;
+
+        public TrackingCallback(final Callback actualCallback) {
+            this.actualCallback = actualCallback;
+        }
+
+        @Override
+        public void onCompletion(final RecordMetadata recordMetadata, final Exception e) {
+            pendingRecords.decrementAndGet();
```

Comment:
Do we want to decrement after the callback is completed? What's the best approach semantically?

Reply:
Since it already happened, we should probably decrement as soon as possible?